test/test.sh

```diff
@@ -0,0 +1,77 @@
+#! /bin/bash
+
+# wgetpaste test script
+# Exit code: number of mismatched downloads or 1 for general failure
+# Copyright (C) 2021  xxc3nsoredxx
+
+# Don't assume the test is being run from the same directory as the script
+TEST_DIR="$(dirname "$0")"
+TEST_FILE="$TEST_DIR/test.txt"
+DL_DIR="$(mktemp -q -d /tmp/wgetpaste_test.XXXXX)"
+DL_COUNT=0
+DL_MISMATCH=0
+
+# Test that temp directory was created
+if [ ! -d "$DL_DIR" ]; then
+    echo "Failed to create temporary download directory: $DL_DIR"
+    exit 1
+fi
+echo "Using download directory: $DL_DIR"
+
+# Post test file into each service (if possible)
+# Download the resulting paste into /tmp/wgetpaste_test.XXXXX/<service>.txt
+for s in $($TEST_DIR/../wgetpaste -S --completions); do
+    # Ignore codepad (timing out)
+    if [ "$s" == 'codepad' ]; then
+        continue
+    fi
+
+    # Discard stderr output
+    echo -n "Posting to $s: "
+    URL="$($TEST_DIR/../wgetpaste -r -s "$s" "$TEST_FILE" 2>/dev/null)"
+    STATUS="$?"
+
+    # Skip failed posts (eg, not authorized for GitHub or GitLab)
+    if [ "$STATUS" -ne 0 ]; then
+        echo "FAILED, skipping..."
```

Comment:
While this is already a nice improvement, it's probably worth reconsidering always skipping here in the future. It makes sense for github, but for example bpaste would just have been skipped as well before 3b7869c2dcfe0fd8e48736a2dcda8dff3f6d5786 landed.

Reply:
Can you elaborate on what you mean? Are you suggesting adding a variable containing a list services to skip and adding GitHub and GitLab to that list? A skip-list could be useful in the case of a generally misbehaving service, but so is testing the `wgetpaste` status. That way the test script doesn't attempt to download from a service if the post failed even though it's not in the list (eg, running the test during server maintenance).

Personally, I don't see a reason to add GitHub and GitLab to said list because the post will fail if the user hasn't set the auth tokens in the config file but _should_ succeed if they have done that.

Reply:
Before I write a long reply, let some quickly drawn up code speak. Makes the difference between the case of needs auth tokens and the paste bin service changed API/URLs/etc more apparent.
```
diff --git a/test/test.sh b/test/test.sh
index ddc25ab..d7e8934 100755
--- a/test/test.sh
+++ b/test/test.sh
@@ -10,6 +10,7 @@ TEST_FILE="$TEST_DIR/test.txt"
 DL_DIR="$(mktemp -q -d /tmp/wgetpaste_test.XXXXX)"
 DL_COUNT=0
 DL_MISMATCH=0
+SKIPPED_COUNT=0
 
 # Test that temp directory was created
 if [ ! -d "$DL_DIR" ]; then
@@ -26,17 +27,24 @@ for s in $($TEST_DIR/../wgetpaste -S --completions); do
         continue
     fi
 
-    # Discard stderr output
     echo -n "Posting to $s: "
-    URL="$($TEST_DIR/../wgetpaste -r -s "$s" "$TEST_FILE" 2>/dev/null)"
+    ERROR_LOG="$DL_DIR"/$s-error.log
+    URL="$($TEST_DIR/../wgetpaste -r -s "$s" --verbose "$TEST_FILE" 2>"$ERROR_LOG")"
     STATUS="$?"
 
     # Skip failed posts (eg, not authorized for GitHub or GitLab)
     if [ "$STATUS" -ne 0 ]; then
-        echo "FAILED, skipping..."
+        if (grep "HTTP.*401.*Unauthorized" "$ERROR_LOG"); then
+            echo "SKIPPING, $s needs authorization"
+            SKIPPED_COUNT=$((SKIPPED_COUNT + 1))
+        else
+            echo "FAILED"
+        fi
+        rm "$ERROR_LOG"
         continue
     fi
     echo "SUCCESS!"
+    rm "$ERROR_LOG"
 
     echo -n "Downloading from $s: "
     if ! (wget -q "$URL" -O "$DL_DIR/$s.txt" 2>/dev/null); then
@@ -67,6 +75,7 @@ for f in $DL_DIR/*; do
 done
 
 echo "Total mismatches: $DL_MISMATCH"
+echo "Skipped tests: $SKIPPED_COUNT"
 
 # Delete download directory if all tests succeeded
 if [ "$DL_MISMATCH" -eq 0 ]; then
```

Reply:
I like that. I split them up into 3 categories of skips: hard skip, auth skip, general failure skip.